pgml-cms/blog/fire.md

```diff
@@ -0,0 +1,228 @@
+
+```python
+async def do_rag(user_query):
+    results = await collection.rag(
+        {
+            "CONTEXT": {
+                "vector_search": {
+                    "query": {
+                        "fields": {
+                            "markdown": {
+                                "query": user_query,
+                                "parameters": {
+                                    "prompt": "Represent this sentence for searching relevant passages: "
+                                },
+                            }
+                        },
+                    },
+                    "document": {"keys": ["id"]},
+                    "rerank": {
+                        "model": "mixedbread-ai/mxbai-rerank-base-v1",
+                        "query": user_query,
+                        "num_documents_to_rerank": 100,
+                    },
+                    "limit": 5,
+                },
+                "aggregate": {"join": "\n\n\n"},
+            },
+            "chat": {
+                "model": "meta-llama/Meta-Llama-3.1-405B-Instruct",
+                "messages": [
+                    {
+                        "role": "system",
+                        "content": "You are a question and answering bot. Answer the users question given the context succinctly.",
+                    },
+                    {
+                        "role": "user",
+                        "content": f"Given the context\n<context>\n:{{CONTEXT}}\n</context>\nAnswer the question: {user_query}",
+                    },
+                ],
+                "max_tokens": 256,
+            },
+        },
+        pipeline,
+    )
+    return results
+```
+
+This function combines vector search, reranking, and text generation to provide context-aware answers to user queries. It uses the Meta-Llama-3.1-405B-Instruct model for text generation.
+
+This query can be broken down into 4 steps:
+1. Perform vector search finding the 100 best matching chunks for the `user_query`
+2. Rerank the results of the vector search using the `mixedbread-ai/mxbai-rerank-base-v1` cross-encoder and limit the results to 5
+3. Join the reranked results with `\n\n\n` and substitute them in place of the `{{CONTEXT}}` placeholder in the messages
+4. Perform text-generation with `meta-llama/Meta-Llama-3.1-405B-Instruct`
+
+This is a complex query and there are more options and parameters to be tuned. See the [Korvus guide to RAG](https://postgresml.org/docs/open-source/korvus/guides/rag) for more information on the `rag` method.
+
+### All Together Now
+
+To tie everything together, we use an interactive loop in our `main()` function:
+
+```python
+async def main():
+    # ... (previous code for setup and indexing)
+
+    # Now we can search
+    while True:
+        user_query = input("\n\nquery > ")
+        if user_query == "q":
+            break
+        results = await do_rag(user_query)
+        print(results)
```

Comment:
Good opportunity to show off some fun example results, and compare it to llama 3.1 w/o RAG.